pkg/migration-tool/writer/shards.go

```diff
@@ -0,0 +1,178 @@
+package writer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/migration-tool/utils"
+)
+
+// shardsSet represents a set of shards. It consists of configuration related
+// to handling and management of shards.
+type shardsSet struct {
+	num         int
+	set         []*shard
+	cancelFuncs []context.CancelFunc
+
+	done    chan struct{}
+	errChan chan error
+}
+
+// newShardsSet creates a shards-set and initializes it. It creates independent clients for each shard,
+// contexts and starts the shards. The shards listens to the writer routine, which is responsible for
+// feeding the shards with data blocks for faster (due to sharded data) flushing.
+func newShardsSet(writerCtx context.Context, writeURL string, numShards int) (*shardsSet, error) {
+	var (
+		set         = make([]*shard, numShards)
+		cancelFuncs = make([]context.CancelFunc, numShards)
+	)
+	ss := &shardsSet{
+		num:     numShards,
+		done:    make(chan struct{}, numShards),
+		errChan: make(chan error, numShards),
+	}
+	for i := 0; i < numShards; i++ {
+		client, err := utils.NewClient(fmt.Sprintf("writer-shard-%d", i), writeURL, utils.Write, model.Duration(defaultWriteTimeout))
+		if err != nil {
+			return nil, fmt.Errorf("creating write-shard-client-%d: %w", i, err)
+		}
+		ctx, cancelFunc := context.WithCancel(writerCtx)
+		shard := &shard{
+			ctx:          ctx,
+			client:       client,
+			queue:        make(chan *[]prompb.TimeSeries),
+			copyShardSet: ss,
+		}
+		cancelFuncs[i] = cancelFunc
+		set[i] = shard
+	}
+	ss.set = set
+	ss.cancelFuncs = cancelFuncs
+	// Run the shards.
+	for i := 0; i < ss.num; i++ {
+		go ss.set[i].run(i)
+	}
+	return ss, nil
+}
+
+// scheduleTS batches the time-series based on the available shards. These batches are then
+// fed to the shards in a single go which then consume and push to the respective clients.
+func (s *shardsSet) scheduleTS(ts []prompb.TimeSeries) {
```

Comment:
I think there needs to be special handling for the progress metric. Right now the progress metric is written by just one of the shards. But what happens if that shard succeeds and updates the progress metric, while the process crashes before the other shards finished? That means some data will be missing on restart, which is not acceptable.

You can look at how we respond to writeRequests in promscale for one idea of how to fix this using waitgroups.

Reply:
I see one way, to send the progress metric in a different POST request after the shards have completed pushing the data.